### PR TITLE
kvserver: abstract away system config span usage in kv      

### PR DESCRIPTION
**Pull request by irfansharif**
Part of #67679. We'll hide `config.SystemConfig` behind the
`spanconfig.StoreReader` interface, and use that instead in the various
queues that need access to the system config span. In future PRs we'll
introduce a data structure that maintains a mapping between spans and
configs that implements this same interface, except it will be powered
by a view over `system.span_configurations` that was introduced in
\#69047.

When we do make that switch, i.e. have KV consult this new thing for
splits, merges, GC and replication, instead of the gossip backed system
config span, ideally it'd be as easy as swapping the source. This PR
helps pave the way for just that.

In #66348 we described how `zonepb.ZoneConfigs` going forward were going
to be an exclusively SQL-level construct. Consequently we purge[*] all
usages of it in KV, storing on each replica a `roachpb.SpanConfig`
instead.

[*]: The only remaining use is what powers our replication reports,
which does not extend well to multi-tenancy in its current form and
needs replacing.

Release note: None

Release justification: low risk, high benefit changes to existing functionality

### CODE REVIEW

**Comment by irfansharif** [comment]
Actually the last commit is reviewable as is.

File: pkg/kv/kvserver/store.go [hunk 1]
```diff
@@ -1734,6 +1735,71 @@ func (s *Store) startGossip() {
...
+var errSysCfgUnavailable = errors.New("system config not available in gossip")
+
+type shadowConfReader struct {
+	new, old spanconfig.StoreReader
+}
+
+func (s *shadowConfReader) NeedsSplit(ctx context.Context, start, end roachpb.RKey) bool {
+	new := s.new.NeedsSplit(ctx, start, end)
+	old := s.old.NeedsSplit(ctx, start, end)
+	if new != old && log.V(1) {
```
Comment (ajwerner) [inline pkg/kv/kvserver/store.go:1747]: log.ExpensiveLogEnabled(ctx, 1)

File: pkg/kv/kvserver/store.go [hunk 2]
```diff
@@ -1734,6 +1735,71 @@ func (s *Store) startGossip() {
 	}
 }
 
+var errSysCfgUnavailable = errors.New("system config not available in gossip")
+
+type shadowConfReader struct {
+	new, old spanconfig.StoreReader
+}
```
Comment (ajwerner) [inline pkg/kv/kvserver/store.go:1742]: please pull this out into a separate file

File: pkg/kv/kvserver/store.go [hunk 3]
```diff
@@ -1734,6 +1735,71 @@ func (s *Store) startGossip() {
...
+	ctx context.Context, key roachpb.RKey,
+) (roachpb.SpanConfig, error) {
+	new, errNew := s.new.GetSpanConfigForKey(ctx, key)
+	old, errOld := s.old.GetSpanConfigForKey(ctx, key)
+	if !new.Equal(old) {
+		log.Warningf(ctx, "get span: mismatched responses new=%s old=%s for key=%s", new.String(), old.String(), key.String())
+	}
+	if !errors.Is(errNew, errOld) {
+		log.Warningf(ctx, "get span: mismatched errors new=%s old=%s for key=%s", errNew, errOld, key.String())
+	}
```
Comment (ajwerner) [inline pkg/kv/kvserver/store.go:1783]: Seems like this is going to be very loud.

File: pkg/kv/kvserver/store.go [hunk 4]
```diff
@@ -1734,6 +1735,71 @@ func (s *Store) startGossip() {
 	}
 }
 
+var errSysCfgUnavailable = errors.New("system config not available in gossip")
+
+type shadowConfReader struct {
```
Comment (ajwerner) [inline pkg/kv/kvserver/store.go:1740]: this deserves commentary. 

It's not obvious that this thing belongs in this package. If it does, it definitely needs a more specific name that references spanconfig

**Review by ajwerner — COMMENTED** [review]
Most of this change feels mechanical. When you can, it's nice to separate out mechanical changes from implementation changes to help the reviewer. 

I don't think this is a particularly controversial change. I'd like Nathan to take a pass as it's the KV team which will end up owning this stuff.

**Comment by irfansharif** [comment]
Woops, the shadowConfReader was not meant to be in this commit (it wasn't wired up to anything anyway -- we'll use it in a later PR when we have another implementation of spanconfig.StoreReader). @nvanbenschoten, mind taking a look at the last commit when you get the chance? It's entirely mechanical.

**Review by nvb — APPROVED** [review]
> Reviewed 59 of 66 files at r7, 77 of 77 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @irfansharif)

___
*[pkg/kv/kvserver/allocator_scorer_test.go, line 620 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2ycMl-I95ffVTDCl1:-Mi2ycMl-I95ffVTDCl2:bjskhda) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e949fa60d8a91eaff0a79a031841362366e2c72/pkg/kv/kvserver/allocator_scorer_test.go#L620)):*
> ```Go
> 				{
> 					Constraints: []zonepb.Constraint{
> 						{Value: "a", Type: zonepb.Constraint_DEPRECATED_POSITIVE},
> ```

This raised a red flag for me and I now see that `Constraint_DEPRECATED_POSITIVE` is not handled in `ZoneConfig.toSpanConfig`. Should it be? Do we have a guarantee that we will never see an existing zone with a `Constraint_DEPRECATED_POSITIVE` constraint type?
___
*[pkg/kv/kvserver/merge_queue.go, line 389 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2dYaIEjni2li_mASJ:-Mi2dYaIEjni2li_mASK:b-y3inik) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/kv/kvserver/merge_queue.go#L389)):*
> ```Go
> 	}
> 	if testingAggressiveConsistencyChecks {
> 		if _, err := mq.store.consistencyQueue.process(ctx, lhsRepl, nil); err != nil {
> ```

Are we intentionally passing nil here?
___
*[pkg/kv/kvserver/replicate_queue.go, line 322 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2iMQ26F_Ed0-lNOsl:-Mi2iMQ26F_Ed0-lNOsm:b-rt324l) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/kv/kvserver/replicate_queue.go#L322)):*
> ```Go
> 
> 			if testingAggressiveConsistencyChecks {
> 				if _, err := rq.store.consistencyQueue.process(ctx, repl, nil); err != nil {
> ```

Same question here. Is this intentional?
___
*[pkg/kv/kvserver/store.go, line 2982 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2xYel4W6fSqsb5fVK:-Mi2xYel4W6fSqsb5fVL:b-nx3jtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/kv/kvserver/store.go#L2982)):*
> ```Go
> }
> 
> // TestingDefaultSpanConfig // XXX:
> ```

Were you planning to remove/rewrite this?
___
*[pkg/roachpb/span_config.proto, line 92 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2esea3AJ-ef3-dW19:-Mi2esea3AJ-ef3-dW1A:ba5rht6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/roachpb/span_config.proto#L92)):*
> ```protobuf
>   // GCTTL is the number of seconds overwritten values will be retained before
>   // garbage collection. A value <= 0 means older versions are never GC-ed.
>   int32 gc_ttl = 3 [(gogoproto.customname) = "GCTTL"];
> ```

Out of curiosity, why did we decide to inline this instead of keeping it split out in a separate policy struct? It feel like we may have lost some future-proofing with that decision, as extensions to the GC policy along multiple dimensions (e.g. retain all deletion tombstones for incremental backup, but eagerly collect duplicate values) will now be more invasive. I wonder how @ajwerner feels about this.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/roachpb/span_config.proto, line 92 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2esea3AJ-ef3-dW19:-Mi3zWJrAHze2rS29eN1:b-nzr1vq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/roachpb/span_config.proto#L92)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Out of curiosity, why did we decide to inline this instead of keeping it split out in a separate policy struct? It feel like we may have lost some future-proofing with that decision, as extensions to the GC policy along multiple dimensions (e.g. retain all deletion tombstones for incremental backup, but eagerly collect duplicate values) will now be more invasive. I wonder how @ajwerner feels about this.
</blockquote></details>

In some ways this feels like the most invasive part of this entire change. I also did not see the motivation. I let it go because I didn't see any strong reason to oppose it. I think your point about policy changes in the future is sound.

**Review by irfansharif — COMMENTED** [review]
TFTRs! Will merge on green.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-:-Mi7d1CJ5ao_jEoYFFnP:b-4jqz1p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/kv/kvserver/allocator_scorer_test.go, line 620 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2ycMl-I95ffVTDCl1:-Mi78O_2Dvzk3ZlDw9CM:b824ui1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e949fa60d8a91eaff0a79a031841362366e2c72/pkg/kv/kvserver/allocator_scorer_test.go#L620)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This raised a red flag for me and I now see that `Constraint_DEPRECATED_POSITIVE` is not handled in `ZoneConfig.toSpanConfig`. Should it be? Do we have a guarantee that we will never see an existing zone with a `Constraint_DEPRECATED_POSITIVE` constraint type?
</blockquote></details>

We're guaranteed, yes. We validate against this when users set zone configs. The only remaining uses of this were in tests.
 https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/config/zonepb/zone.go#L368-L371

Even looking back to the change that deprecated it, it seems they were never officially supported or documented.
https://github.com/cockroachdb/cockroach/commit/73466fd3eff5e42667c0ca9bfc0f98bdbcb0ec96

I suppose it's possible for some on-prem customer running v1.x to have this lying around, and an upgrade to 21.2 would break it cause we never blocked the upgrade with this validation. Do we care about that at all?
___
*[pkg/kv/kvserver/merge_queue.go, line 389 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2dYaIEjni2li_mASJ:-Mi7BjPT-XZEUaA00HT1:b-wc56ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/kv/kvserver/merge_queue.go#L389)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are we intentionally passing nil here?
</blockquote></details>

It was, we passed in nil elsewhere for queues that don't depend on the system config span, and these queues don't either, but meh I'll pass in the available conf reader.

https://github.com/cockroachdb/cockroach/blob/56afa9f4afb8211ef7783b7069cded1e1a880bed/pkg/kv/kvserver/store.go#L2879
___
*[pkg/kv/kvserver/replicate_queue.go, line 322 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2iMQ26F_Ed0-lNOsl:-Mi7DRafBbc1M9w4mvos:b-q64pkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/kv/kvserver/replicate_queue.go#L322)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same question here. Is this intentional?
</blockquote></details>

See above.
___
*[pkg/kv/kvserver/store.go, line 2982 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2xYel4W6fSqsb5fVK:-Mi7DVsH0wOUyAB36y69:be9o5be) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/kv/kvserver/store.go#L2982)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Were you planning to remove/rewrite this?
</blockquote></details>

No, this will be here for now, but I did forget to add a proper comment. Done.
___
*[pkg/roachpb/span_config.proto, line 92 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-Mi2esea3AJ-ef3-dW19:-Mi7NGhqDI1qng5dZ3aU:b-mbarlr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296bac14efcd7228fe323d23fe99dca6c3804bc/pkg/roachpb/span_config.proto#L92)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

In some ways this feels like the most invasive part of this entire change. I also did not see the motivation. I let it go because I didn't see any strong reason to oppose it. I think your point about policy changes in the future is sound. 
</blockquote></details>

Hm, I was inclined to simplify it given the struct in its original form was introduced with the single field back in 2014 and hadn't seen any change since. But I don't feel strongly about it either way -- changed to a similar looking policy struct.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 8 of 10 files at r9, 2 of 2 files at r10, 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69172#-:-MiMTr4h85nPP06FhjX1:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

**Pull request merged**